cpp/src/knn/knn_opg_common.cuh

```diff
@@ -1,137 +1,916 @@
+struct KNN_RE_params : public opg_knn_param<in_t, ind_t, dist_t, out_t> {
+  KNN_RE_params(knn_operation knn_op,
+                std::vector<Matrix::Data<in_t> *> *idx_data,
+                Matrix::PartDescriptor *idx_desc,
+                std::vector<Matrix::Data<in_t> *> *query_data,
+                Matrix::PartDescriptor *query_desc, bool rowMajorIndex,
+                bool rowMajorQuery, size_t k, size_t batch_size, bool verbose,
+                int n_outputs, std::vector<std::vector<out_t *>> *y,
+                std::vector<Matrix::Data<out_t> *> *out)
+    : opg_knn_param<in_t, ind_t, dist_t, out_t>(
+        knn_op, idx_data, idx_desc, query_data, query_desc, rowMajorIndex,
+        rowMajorQuery, k, batch_size, verbose) {
+    this->n_outputs = n_outputs;
+    this->y = y;
+    this->out = out;
+  }
+};
+
+template <typename in_t, typename ind_t, typename dist_t, typename out_t>
+struct KNN_CL_params : public opg_knn_param<in_t, ind_t, dist_t, out_t> {
+  KNN_CL_params(knn_operation knn_op,
+                std::vector<Matrix::Data<in_t> *> *idx_data,
+                Matrix::PartDescriptor *idx_desc,
+                std::vector<Matrix::Data<in_t> *> *query_data,
+                Matrix::PartDescriptor *query_desc, bool rowMajorIndex,
+                bool rowMajorQuery, size_t k, size_t batch_size, bool verbose,
+                int n_outputs, std::vector<std::vector<out_t *>> *y,
+                std::vector<int> *n_unique, std::vector<out_t *> *uniq_labels,
+                std::vector<Matrix::Data<out_t> *> *out,
+                std::vector<std::vector<float *>> *probas)
+    : opg_knn_param<in_t, ind_t, dist_t, out_t>(
+        knn_op, idx_data, idx_desc, query_data, query_desc, rowMajorIndex,
+        rowMajorQuery, k, batch_size, verbose) {
+    this->n_outputs = n_outputs;
+    this->y = y;
+    this->n_unique = n_unique;
+    this->uniq_labels = uniq_labels;
+    this->out = out;
+    this->probas = probas;
+  }
+};
+
+/**
+ * A structure to store utilities for CUDA and RAFT comms
+ */
+struct cuda_utils {
+  cuda_utils(raft::handle_t &handle) {
+    this->alloc = handle.get_device_allocator();
+    this->stream = handle.get_stream();
+    this->comm = &(handle.get_comms());  //communicator_ is a private attribute
+    size_t n_internal_streams = handle.get_num_internal_streams();
+    this->internal_streams.resize(n_internal_streams);
+    for (int i = 0; i < n_internal_streams; i++) {
+      internal_streams[i] = handle.get_internal_stream(i);
+    }
+  }
+  std::shared_ptr<deviceAllocator> alloc; /**< RMM alloc */
+  cudaStream_t stream;                    /**< CUDA user stream */
+  const raft::comms::comms_t *comm;       /**< RAFT comms handle */
+  std::vector<cudaStream_t>
+    internal_streams; /**< Vector of CUDA internal streams */
+};
+
+/**
+ * A structure to store utilities for distributed KNN operations
+ */
+template <typename in_t, typename ind_t, typename dist_t, typename out_t>
+struct opg_knn_work {
```

Comment:
We don't need to hold up this PR for such a small nitpick, but the `_work` suffix feels like this should actually be doing some processing, rather than just holding state about the workers on the system. Something like `opg_knn_state` might be more fitting.